### DISCUSSION TITLE
Custom username

### DISCUSSION THREAD

**Discussion by ptyork** [community/community#43591] [Codespaces] [answered]
### Select Topic Area

Question

### Body

I'm banging my head on this one. I'm trying to have the codespace environment mimic as close as possible the environment a student would see in a "real world" environment. And make their assignments as "uncopypasteable" as possible. As such, I'm trying to get their usernames to match their github usernames.

I've gotten this to work using the [Common Utils](https://github.com/devcontainers/features/tree/main/src/common-utils) feature and setting the username to ${localEnv:GITHUB_USER} and then further setting remoteUser to the same. BUT, this doesn't work if you prebuild the container, as GITHUB_USER is not set during prebuild and the "features" are baked into the images.

So I thought to use one of the [lifecycle scripts](https://containers.dev/implementors/json_reference/#lifecycle-scripts). But these all run as the user you have set as "remoteUser". Chicken and egg. I CAN create the user in an onCreateCommand, but I can't then connect as that user.

Can you think of any way to prebuild AND connect as a customized user?

**Comment by cwndrws** [comment, marked as answer]
Hi! Thank you for writing in!

You are correct that your current setup using features will only work without prebuilds since you're prebuilding in order to mitigate the wait on the features install for the user. Are you just trying to set the username to have it shown in the bash prompt? Or do you actually want a user on the box with that name?

Also, are you using the default image? Or providing a custom image?

**Reply by ptyork** [reply]
I'm constructing a custom image with MariaDB, Python, and a few other tools for a Database course. I mostly want the custom user so that I can have students work in MariaDB with their own usernames, set passwords, etc. Just a more realistic experience.

I have basically abandoned the pre-build concept at this point because the prebuilds don't seem to survive the repo cloning process that is part of the GitHub Classroom workflow. Instead I have created a custom `non-root-user` feature by stripping out almost everything from `common-utils` except for the relevant pieces. This seems to have made the build time much more tolerable (down to a minute or so from five). Still working through some custom rc file fun, but I think I may have found an acceptable solution.

That said, the only part of the container build that is in the json file is the custom feature. Dockerfile contains everything else. If there's a way to cache the raw docker image (from the docker build) within an organization (classroom) and between repos, that could save a bit more time (and server-side resources).

Thanks!

**Reply by cwndrws** [reply]
I'm glad you got your start time down, but I'm sorry prebuilds didn't work out.

If you're familiar enough with Docker, generally, you could build the image and store it in a registry to be pulled. You can specify an image to pull rather than a Dockerfile to build in your `devcontainer.json` https://containers.dev/implementors/json_reference/#image-specific. That obviously is more setup than you require now and I'm not sure it would get improve start time drastically enough to be worth it, but it's something to consider.

Best of luck!

**Reply by ptyork** [reply]
Thanks! Certainly worth a try. You're likely right that the one-time build overhead isn't likely worth the headache, but I've been needing to look into distributing docker images from Docker Hub for a different project, anyway. So two birds...
